src/consensus/mod.rs

```diff
@@ -1,58 +1,191 @@
         }
     }
 
-    pub async fn propose(&self, proposal: Vec<Endpoint>) {
-        // TODO: register proposal with regular paxos instance
-        unimplemented!();
+    /// Propose a new membership set to the cluster
+    ///
+    /// # Errors
+    ///
+    /// Returns `NewBrokenPipe` if the broadcast was not sucessful
+    pub async fn propose(&mut self, proposal: Vec<Endpoint>) -> Result<()> {
+        let mut paxos_delay = Delay::new(Instant::now() + self.get_random_delay()).fuse();
+
+        async {
+            paxos_delay.await;
+            self.start_classic_round().await;
+        }
+            .await;
+
+        let (tx, rx) = oneshot::channel();
+
+        let request = Request::new_fast_round(tx, self.my_addr.clone(), self.config_id, proposal);
+
+        // TODO: Handle Vec<Result<Response>>
+        self.broadcast.broadcast(request).await;
+        rx.await.map_err(|_| Error::new_broken_pipe(None))?;
+
+        Ok(())
     }
 
-    pub async fn handle_message(&self, request: Request) -> Result<Response> {
+    /// Handles a consensus message which the membership service reveives.
+    ///
+    /// # Errors
+    ///
+    /// * `NewUnexpectedRequest`: if any other request kind other tha a Consensus message is
+    /// passed.
+    /// * `VoteAlredyReceived`: If the request is from a sender who has already cast a vote which
+    /// this node as received.
+    /// * `AlreadyReachedConsensus`: If this is called after the cluster has already reached
+    /// consensus.
+    /// * `FastRoundFailure`: If fast paxos is unable to reach consensus
+    pub async fn handle_message(&mut self, request: Request) -> Result<Vec<Endpoint>> {
         match request.kind() {
-            proto::RequestKind::Consensus(_) => self.handle_fast_round(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_1a(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_1b(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_2a(request).await,
-            proto::RequestKind::Consensus(_) => self.paxos.handle_phase_2b(request).await,
+            proto::RequestKind::Consensus(req_type) => {
+                if self.paxos.is_some() {
+                    match req_type {
+                        proto::Consensus::FastRoundPhase2bMessage(msg) => {
+                            self.handle_fast_round(msg).await
+                        }
+                        // TODO: need to figure out communication b/w paxos and fast paxos
+                        _ => unimplemented!(),
+                    }
+                } else {
+                    Err(Error::new_unexpected_request(None))
+                }
+            }
             _ => Err(Error::new_unexpected_request(None)),
         }
     }
 
-    async fn handle_fast_round(&self, request: Request) -> crate::Result<Response> {
-        unimplemented!()
+    async fn handle_fast_round<'b>(
```

Comment:
can we check if we can drop these lifetimes on the latest nightly? I think that pr got merged